File: src/agent/providers/anthropic.ts
import { CanonicalMessage, Provider, ProviderRequest, ProviderStreamEvent, ToolCall, ToolSchema } from "../types";
import { FetchLike, openProviderStream, readSseEvents, sseData, toTransportError, usageEvent } from "./transport";

const ANTHROPIC_URL = "https://api.anthropic.com/v1/messages";
const ANTHROPIC_VERSION = "2023-06-01";
/** Required by the API, unlike OpenAI's where it's optional. Generous: a rewrite of a large note is the worst case. */
const MAX_TOKENS = 8192;

export type AnthropicProviderOptions = {
  apiKey: string;
  model: string;
  /** See OpenRouterProviderOptions — must be `expo/fetch` in the app. */
  fetch?: FetchLike;
};

/**
 * Anthropic's Messages API. A genuinely different protocol from OpenAI's, in
 * four ways that all matter here:
 *
 * - The system prompt is a **top-level field**, not a message with role `system`.
 * - Tool calls and their results are **content blocks** (`tool_use` /
 *   `tool_result`) inside user and assistant messages, not a separate `tool` role.
 * - Tool schemas use `input_schema`, not `parameters`.
 * - Streaming is **named events** (`content_block_delta` and friends) rather than
 *   one chunk shape, and tool arguments arrive as `partial_json` fragments that
 *   have to be accumulated per block index.
 *
 * Also note the auth header is `x-api-key`, not `Authorization: Bearer`.
 */
export class AnthropicProvider implements Provider {
  readonly id = "anthropic";

  constructor(private opts: AnthropicProviderOptions) {}

  async *send(req: ProviderRequest): AsyncIterable<ProviderStreamEvent> {
    const body = {
      model: this.opts.model,
      max_tokens: MAX_TOKENS,
      stream: true,
      system: req.system,
      messages: toAnthropicMessages(req.messages),
      ...(req.tools.length > 0 ? { tools: req.tools.map(toAnthropicTool) } : {}),
    };

    const opened = await openProviderStream({
      url: ANTHROPIC_URL,
      displayName: "Anthropic",
      fetch: this.opts.fetch,
      init: {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "x-api-key": this.opts.apiKey,
          "anthropic-version": ANTHROPIC_VERSION,
        },
        body: JSON.stringify(body),
        signal: req.signal,
      },
    });

    if ("error" in opened) {
      yield { type: "error", error: opened.error };
      return;
    }

    // Aborting mid-stream rejects the reader, and `readSseEvents` deliberately
    // lets that through so it can be classified here alongside every other
    // transport failure rather than escaping `runTurn` as a raw exception.
    try {
      yield* parseAnthropicStream(opened.stream);
    } catch (err) {
      yield { type: "error", error: toTransportError(err, req.signal) };
    }
  }
}

/** Tool blocks are keyed by index while their arguments stream in as text fragments. */
type ToolBlock = { id: string; name: string; json: string };

async function* parseAnthropicStream(stream: ReadableStream<Uint8Array>): AsyncIterable<ProviderStreamEvent> {
  const blocks = new Map<number, ToolBlock>();

  for await (const event of readSseEvents(stream)) {
    const data = sseData(event);
    if (data === null) continue;

    let parsed: AnthropicEvent;
    try {
      parsed = JSON.parse(data) as AnthropicEvent;
    } catch {
      continue; // a malformed chunk isn't worth failing the turn over
    }

    switch (parsed.type) {
      case "content_block_start": {
        const block = parsed.content_block;
        if (block?.type === "tool_use" && typeof block.id === "string" && typeof block.name === "string") {
          blocks.set(parsed.index ?? 0, { id: block.id, name: block.name, json: "" });
        }
        break;
      }

      case "content_block_delta": {
        const delta = parsed.delta;
        if (delta?.type === "text_delta" && typeof delta.text === "string" && delta.text.length > 0) {
          yield { type: "text", delta: delta.text };
        }
        if (delta?.type === "input_json_delta" && typeof delta.partial_json === "string") {
          const block = blocks.get(parsed.index ?? 0);
          if (block) block.json += delta.partial_json;
        }
        break;
      }

      case "content_block_stop": {
        const index = parsed.index ?? 0;
        const block = blocks.get(index);
        if (block) {
          blocks.delete(index);
          yield { type: "toolCall", call: finishToolBlock(block) };
        }
        break;
      }

      case "message_delta":
      case "message_start": {
        // Input tokens arrive on message_start, output on message_delta's usage.
        const usage = parsed.usage ?? parsed.message?.usage;
        const usageReport = usage ? usageEvent(usage.input_tokens, usage.output_tokens) : null;
        if (usageReport) yield usageReport;
        break;
      }

      case "error": {
        const message = parsed.error?.message;
        yield {
          type: "error",
          error: { kind: "unknown", message: typeof message === "string" ? message : "Anthropic reported an error." },
        };
        break;
      }
    }
  }

  // A stream that ends without content_block_stop still holds complete calls.
  // Same failure mode as the OpenAI adapter's trailing flush: dropping them
  // makes the agent look like it ignored the user.
  for (const block of blocks.values()) {
    yield { type: "toolCall", call: finishToolBlock(block) };
  }
}

function finishToolBlock(block: ToolBlock): ToolCall {
  // Malformed input becomes empty arguments rather than a dropped call, so the
  // tool layer can return a model-facing error the model can correct.
  try {
    const parsed: unknown = block.json.trim().length > 0 ? JSON.parse(block.json) : {};
    const usable = typeof parsed === "object" && parsed !== null && !Array.isArray(parsed);
    return { id: block.id, name: block.name, arguments: usable ? (parsed as Record<string, unknown>) : {} };
  } catch {
    return { id: block.id, name: block.name, arguments: {} };
  }
}

type AnthropicMessage = { role: "user" | "assistant"; content: unknown[] };

/**
 * Appends a content block, merging into the previous message when the role
 * matches.
 *
 * The API rejects two messages with the same role in a row, and the canonical
 * history does not guarantee alternation: a `tool` result becomes a *user*
 * message here, so a turn that ended on one is immediately followed by the next
 * turn's prompt. Merging by role is what makes every history shape sendable
 * rather than only the ones a clean turn produces.
 */
function appendBlock(out: AnthropicMessage[], role: "user" | "assistant", block: unknown): void {
  const last = out[out.length - 1];
  if (last?.role === role) last.content.push(block);
  else out.push({ role, content: [block] });
}

/**
 * Canonical messages → Anthropic's shape.
 *
 * The awkward part is that a `tool` result is not its own role: it must ride
 * inside a **user** message as a `tool_result` block.
 */
function toAnthropicMessages(messages: CanonicalMessage[]): unknown[] {
  const out: AnthropicMessage[] = [];

  for (const m of messages) {
    if (m.role === "tool") {
      appendBlock(out, "user", { type: "tool_result", tool_use_id: m.toolCallId, content: m.content });
      continue;
    }

    if (m.role === "assistant" && "toolCalls" in m) {
      for (const tc of m.toolCalls) {
        appendBlock(out, "assistant", { type: "tool_use", id: tc.id, name: tc.name, input: tc.arguments });
      }
      continue;
    }

    appendBlock(out, m.role, { type: "text", text: (m as { content: string }).content });
  }

  return out;
}

function toAnthropicTool(tool: ToolSchema): unknown {
  // `input_schema`, not `parameters`, and not nested under a `function` key.
  return { name: tool.name, description: tool.description, input_schema: tool.parameters };
}

type AnthropicEvent = {
  type?: string;
  index?: number;
  content_block?: { type?: string; id?: unknown; name?: unknown };
  delta?: { type?: string; text?: unknown; partial_json?: unknown };
  usage?: { input_tokens?: unknown; output_tokens?: unknown };
  message?: { usage?: { input_tokens?: unknown; output_tokens?: unknown } };
  error?: { message?: unknown };
};

File: lib/features/ai/data/agent/providers/anthropic.dart
import 'dart:convert';

import '../../../models/ai_allow_list.dart';
import '../reasoning.dart';
import '../transport.dart';
import '../types.dart';

const _anthropicUrl = 'https://api.anthropic.com/v1/messages';
const _anthropicVersion = '2023-06-01';
const _maxTokens = 8192;

/// Anthropic's Messages API. A genuinely different protocol from OpenAI's:
/// the system prompt is a top-level field, tool calls/results are content
/// blocks (`tool_use` / `tool_result`) inside assistant/user messages, tool
/// schemas use `input_schema`, and streaming is named events with `partial_json`
/// fragments accumulated per block index. Auth is `x-api-key`, not a bearer.
class AnthropicProvider implements Provider {
  AnthropicProvider({
    required this.apiKey,
    required this.model,
    required this.reasoning,
    HttpTransport? transport,
    String displayName = 'Anthropic',
  })  : _transport = transport ?? HttpTransportImpl(),
        _displayName = displayName;

  final String apiKey;
  final String model;
  final ReasoningMode reasoning;
  final HttpTransport _transport;
  final String _displayName;

  @override
  Stream<ProviderStreamEvent> send(ProviderRequest request) async* {
    final body = {
      'model': model,
      'max_tokens': _maxTokens,
      'stream': true,
      'system': request.system,
      'messages': _toAnthropicMessages(request.messages),
      if (request.tools.isNotEmpty)
        'tools': request.tools.map(_toAnthropicTool).toList(growable: false),
      ..._thinking(),
    };

    HttpResponseLike opened;
    try {
      opened = await _transport.post(
        Uri.parse(_anthropicUrl),
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': apiKey,
          'anthropic-version': _anthropicVersion,
        },
        body: jsonEncode(body),
        token: request.cancellation,
      );
    } catch (err) {
      yield ProviderErrorMessage(toTransportError(err, request.cancellation));
      return;
    }

    if (!opened.ok) {
      final raw = await opened.text();
      final message = errorMessageFromBody(
          raw, '$_displayName returned HTTP ${opened.status}.');
      yield ProviderErrorMessage(
        classifiedError(opened.status, message,
            '$_displayName returned HTTP ${opened.status}.'),
      );
      return;
    }

    final stream = opened.body;
    if (stream == null) {
      yield const ProviderErrorMessage(
        ProviderError(
            kind: ProviderErrorKind.unknown,
            message: 'The response did not stream.'),
      );
      return;
    }

    try {
      yield* _parseAnthropicStream(stream, request.cancellation);
    } catch (err) {
      yield ProviderErrorMessage(toTransportError(err, request.cancellation));
    }
  }

  Map<String, dynamic> _thinking() {
    switch (reasoning) {
      case ReasoningMode.disabled:
        return {
          'thinking': {'type': 'disabled'}
        };
      case ReasoningMode.enabled:
        return {
          'thinking': {'type': 'enabled', 'budget_tokens': 4096}
        };
      default:
        final effort = effortValue(reasoning)!;
        // Claude 5 models are adaptive-only (budget ignored) and take the
        // effort level via output_config.effort; budget_tokens is kept for
        // older budget-based models where it is the only knob.
        return {
          'thinking': {'type': 'enabled', 'budget_tokens': 8192},
          'output_config': {'effort': effort},
        };
    }
  }
}

class _ToolBlock {
  _ToolBlock({required this.id, required this.name});
  final String id;
  final String name;
  String json = '';
}

Stream<ProviderStreamEvent> _parseAnthropicStream(
    Stream<List<int>> stream, CancellationToken token) async* {
  final blocks = <int, _ToolBlock>{};

  await for (final event in readSseEvents(stream, token: token)) {
    if (event.data.isEmpty) continue;
    Object? parsed;
    try {
      parsed = jsonDecode(event.data);
    } catch (_) {
      continue;
    }
    if (parsed is! Map<String, dynamic>) continue;

    switch (parsed['type']) {
      case 'content_block_start':
        final block = parsed['content_block'];
        if (block is Map<String, dynamic> && block['type'] == 'tool_use') {
          final id = block['id'];
          final name = block['name'];
          final index = parsed['index'];
          if (id is String && name is String) {
            blocks[index is num ? index.toInt() : 0] =
                _ToolBlock(id: id, name: name);
          }
        }

      case 'content_block_delta':
        final delta = parsed['delta'];
        if (delta is! Map<String, dynamic>) continue;
        final text = delta['text'];
        if (delta['type'] == 'text_delta' &&
            text is String &&
            text.isNotEmpty) {
          yield TextDelta(text);
        }
        final partialJson = delta['partial_json'];
        if (delta['type'] == 'input_json_delta' && partialJson is String) {
          final index = parsed['index'];
          final block = blocks[index is num ? index.toInt() : 0];
          if (block != null) block.json += partialJson;
        }

      case 'content_block_stop':
        final index = parsed['index'];
        final block = blocks.remove(index is num ? index.toInt() : 0);
        if (block != null) yield ToolCallEvent(_finishToolBlock(block));

      case 'message_delta':
      case 'message_start':
        final usage = parsed['usage'] ?? parsed['message'];
        if (usage is Map<String, dynamic>) {
          final report = usageEvent(
            usage['input_tokens'] is num
                ? (usage['input_tokens'] as num).toInt()
                : null,
            usage['output_tokens'] is num
                ? (usage['output_tokens'] as num).toInt()
                : null,
          );
          if (report != null) yield report;
        }

      case 'error':
        final error = parsed['error'];
        final message =
            error is Map<String, dynamic> && error['message'] is String
                ? error['message'] as String
                : 'Anthropic reported an error.';
        yield ProviderErrorMessage(
            ProviderError(kind: ProviderErrorKind.unknown, message: message));
    }
  }

  // A stream that ended without content_block_stop still holds complete calls.
  for (final block in blocks.values) {
    yield ToolCallEvent(_finishToolBlock(block));
  }
}

ToolCall _finishToolBlock(_ToolBlock block) {
  return ToolCall(
    id: block.id,
    name: block.name,
    arguments: _parseArguments(block.json),
  );
}

Map<String, dynamic> _parseArguments(String raw) {
  if (raw.trim().isEmpty) return const {};
  try {
    final parsed = jsonDecode(raw);
    if (parsed is Map<String, dynamic>) return parsed;
  } catch (_) {
    // fall through
  }
  return const {};
}

/// Anthropic rejects two messages with the same role in a row, so blocks are
/// merged into the previous message when the role matches. A `tool` result
/// rides inside a **user** message as a `tool_result` block.
List<Map<String, dynamic>> _toAnthropicMessages(
    List<CanonicalMessage> messages) {
  final out = <Map<String, dynamic>>[];

  void appendBlock(String role, Object block) {
    final last = out.isEmpty ? null : out.last;
    if (last != null && last['role'] == role && last['content'] is List) {
      (last['content'] as List).add(block);
    } else {
      out.add({
        'role': role,
        'content': [block]
      });
    }
  }

  for (final message in messages) {
    switch (message) {
      case ToolResultMessage(:final toolCallId, :final content):
        appendBlock('user', {
          'type': 'tool_result',
          'tool_use_id': toolCallId,
          'content': content
        });

      case AssistantToolCalls(:final toolCalls):
        for (final call in toolCalls) {
          appendBlock('assistant', {
            'type': 'tool_use',
            'id': call.id,
            'name': call.name,
            'input': call.arguments
          });
        }

      case UserMessage(:final content):
        appendBlock('user', {'type': 'text', 'text': content});
      case AssistantMessage(:final content):
        appendBlock('assistant', {'type': 'text', 'text': content});
    }
  }

  return out;
}

/// `input_schema`, not `parameters`, and not nested under a `function` key.
Map<String, dynamic> _toAnthropicTool(ToolSchema tool) => {
      'name': tool.name,
      'description': tool.description,
      'input_schema': tool.parameters,
    };
